sklearn/impute.py

```diff
@@ -891,14 +908,16 @@ def fit_transform(self, X, y=None):
 
         abs_corr_mat = self._get_abs_corr_mat(Xt)
 
-        # impute data
         n_samples, n_features = Xt.shape
         self.imputation_sequence_ = []
         if self.verbose > 0:
             print("[IterativeImputer] Completing matrix with shape %s"
                   % (X.shape,))
         start_t = time()
-        for i_rnd in range(self.n_iter):
+        self.n_iter_ = self.max_iter
+        if not self.sample_posterior:
+            Xt_previous = np.copy(Xt)
+        for i_rnd in range(self.max_iter):
```

Comment:
I would prefer a `while` loop

```python
while criterion > self.tol or self.n_iter_ < self.max_iter
```

The reasoning is that we already see that we will have an early stopping mechanism

Reply:
You also need `if not self.sample_posterior:` so it gets messier. I would prefer to leave as is.

Reply:
The idiom I have come to enjoy is:
```py
for self.n_iter_ in range(1, self.max_iter + 1):
    update model

    if stopping condition is met:
        break
else:
    warnings.warng(...)
```

(We have something similar in label_propagation.py, but there the stopping condition applies at the start of the loop)